### DISCUSSION TITLE
Reg: Generating 3D columnar hexagonal grain

### DISCUSSION THREAD

**Discussion by sbmse007** [idaholab/moose#25155] [Q&A Modules: Phase field]
Hello,
I have been trying to generate 3D columnar hexagonal grains. I followed the "hex_grain_growth_2D_eldrforce.i" input file. However, when I am converting the file into 3D, I am always getting the following error:
```
"*** ERROR ***
The following error occurred in the object "hex_ic", of type "PolycrystalHex".
PolycrystalHex requires a square or cubic number depending on the mesh dimension".
```
In the 2D version, putting op_num = 3 as well as  grain_num = 36 in the '"PolycrystalHex" user object works fine.
Moreover, I can eliminate the above mentioned error if I use op_num = 8 and grain_num = 8. But the grain structure is not a perfect hexagon as in 2D. It will be very helpful if anyone can provide me with some suggestions.

**Comment by hugary1995** [comment]
What does the 3D result look like? Your op_num should be big enough to allow for grain coloring (which is np hard)... 8 is typically good for 3D columnar grains.

**Reply by GiudGiud** [reply]
These ics are relying on the underlying mesh to represent the shapes right? So to have perfect hexagons you would need a very regular TRI mesh and the IC to perfectly align?

**Reply by sbmse007** [reply]
Hi, 
I am attaching the input file as well as the generated output in the attachment. I know I am missing something, but not able to figure out.
![Hex_pseudo3D](https://github.com/idaholab/moose/assets/141843658/b37985d8-a54a-4cdf-8e7e-49b95213883d)
[hex_3D.txt](https://github.com/idaholab/moose/files/12304447/hex_3D.txt)

**Reply by GiudGiud** [reply]
Can you try without using uniform refinement?
I think Ive seen this problem on another IC

**Reply by sbmse007** [reply]
Hi,
I am attaching the output without refinement. But I am not getting the result as seen in the 2D case. 
![Hex_pseudo3D_without_refine](https://github.com/idaholab/moose/assets/141843658/fd365924-f7c2-4d85-b2e0-f02c75aeaccc)
![Hex_2D](https://github.com/idaholab/moose/assets/141843658/07d5e05c-5d11-41cb-b813-a3e77861c159)

**Reply by hugary1995** [reply]
It seems to me that PolycrystalHex isn't compatible with `columnar_3D = true`. We should check that and error. Please open an issue (or a feature request if you want columnar hexes in 3D).

**Reply by sourabhkadambi** [reply]
Hello! This discussion and issue is similar to the problem I was working on recently. I needed the 3D columnar hexagonal grains, and was able to get it to work (see attached rendering) with minor code changes to `PolycrystalHex`. I can plan to contribute the changes.
![Screenshot 2023-08-09 at 1 55 07 PM](https://github.com/idaholab/moose/assets/74273633/58e1b1d1-b64e-43c4-8c3f-da31509bbc2f)

**Reply by GiudGiud** [reply]
awesome!
please do and tag Gary or myself

**Reply by permcody** [reply]
As a possible author of this code, I'm not sure it ever worked in 3D... You likely aren't doing anything wrong other than hitting half-baked code. @sourabhkadambi - I would welcome your changes to the repository!

**Reply by hugary1995** [reply]
@permcody is right. For now if you make a PR, just error out if mesh is 3D but columnar_3D is false.

**Reply by GiudGiud** [reply]
(unless you implemented the functionality, that s better than an error)

**Comment by sbmse007** [comment]
@sourabhkadambi- This is great. Can you also please provide me with some hints to generate something like yours? I am also presently working on modifying the code to make it work.
Thanks.

**Reply by sourabhkadambi** [reply]
Hello, the PR with `_columnar_3D` option has been merged. You should be able to initialized the required microstructure now.
